rmk/src/keyboard.rs

```diff
@@ -152,255 +143,188 @@ impl<'a, M: MatrixTrait, const ROW: usize, const COL: usize, const NUM_LAYER: us
-            // TODO: make it configurable
-            if cur_tick - self.last_mouse_tick > 20 {
-                sender
-                    .send(KeyboardReportMessage::CompositeReport(
-                        self.other_report,
-                        CompositeReportType::Mouse,
-                    ))
-                    .await;
-                self.last_mouse_tick = cur_tick;
-            }
-            // Do nothing
-            if self.other_report.x == 0
-                && self.other_report.y == 0
-                && self.other_report.wheel == 0
-                && self.other_report.pan == 0
-            {
-                // Release, stop report mouse report
-                self.need_send_mouse_report = false;
-            }
-            yield_now().await;
-        }
+    pub(crate) async fn send_mouse_report(&mut self) {
+        self.sender
+            .send(KeyboardReportMessage::CompositeReport(
+                self.other_report,
+                CompositeReportType::Mouse,
+            ))
+            .await;
     }
 
     /// Main keyboard task, it scans matrix, processes active keys.
-    /// If there is any change of key states, set self.changed=true.
-    ///
-    /// `sender` is required because when there's tap action, the report should be sent immediately and then continue scanning
-    pub(crate) async fn scan_matrix(
-        &mut self,
-        sender: &Sender<'a, CriticalSectionRawMutex, KeyboardReportMessage, 8>,
-    ) {
-        #[cfg(feature = "async_matrix")]
-        self.matrix.wait_for_key().await;
-
-        // Matrix scan
-        self.matrix.scan().await;
-
-        // Check matrix states, process key if there is a key state change
-        // Keys are processed in the following order:
-        // process_key_change -> process_key_action_* -> process_action_*
-        for row_idx in 0..self.matrix.get_row_num() {
-            for col_idx in 0..self.matrix.get_col_num() {
-                let ks = self.matrix.get_key_state(row_idx, col_idx);
-                if ks.changed {
-                    self.process_key_change(row_idx, col_idx, sender).await;
-                } else if ks.pressed {
-                    // When there's no key change, only tap/hold action needs to be processed
-                    // Continuously check the hold state
-                    let action = self
-                        .keymap
-                        .borrow_mut()
-                        .get_action_with_layer_cache(row_idx, col_idx, ks);
-                    let tap_hold = match action {
-                        KeyAction::TapHold(tap_action, hold_action) => {
-                            Some((tap_action, hold_action))
+    pub(crate) async fn scan_matrix(&mut self) {
+        let mut ticker = Ticker::every(Duration::from_millis(1));
+        let mut holds: Vec<Holdingkey, 16> = Vec::new();
+        let mut holding = 0;
+        loop {
```

Comment:
This will limit the scan rate to 1KHZ. I want to improve the scan rate as fast as possible(currently it's set to 10KHZ, but I'll unlock it in the future), so I won't accept this change.

Reply:
Since it's doing more things now, I just set it a conservative value, to make sure don't starve other code.

Yes, should be easy to change to 10KHz. But BLE and USB limit report to every 7ms, I think.

Reply:
The report interval is typically 7.5ms for BLE and 1ms/0.125ms for USB2.0. 

Ideally this ticker should be removed to reach the maximum scanning speed. Then the scan rate can be much larger than actually report rate, ensuring best performance. Since there're lots of .await points, so other tasks won't be starved. 

The potential issue is when the report generating speed is larger than host request speed, then everything blocks. That's what I mentioned as `flooding report`

Reply:
block happens when channel between keyboard and communication task full. That's why I want to increase that one.
Other than occasional burst, caused by releasing key trigger the many previously hold keys, human won't able to type that fast to cause problem.

Reply:
Yeah but it's still possible, when processing a macro or a mouse(trackpad), which may send lots of report continuously. Merging report if possible may help. Increasing the channel size would also help, but it's hard to say what size is good.